src/include/k5-int.h

```diff
@@ -855,6 +861,13 @@ krb5_pa_data *
 krb5int_find_pa_data(krb5_context, krb5_pa_data *const *, krb5_preauthtype);
 /* Does not return a copy; original padata sequence responsible for freeing*/
 
+/* Given data, length and PA type, allocate a pa data and add it to a list.
+ * If called with NULL, a new one-element list is returned.
+ * The data will be owned by the list on success.*/
+krb5_error_code
+k5_add_pa_data_element(krb5_preauthtype, size_t, krb5_octet *,
+                       krb5_pa_data ***);
+
```

Comment:
When I suggested adding this helper, I probably should have considered all of the nitpicky details that would come along with it.

First, since the list is in-out, I normally think of this as as mutator rather than a function with an output parameter, so the list parameter should go first.  All of the existing helpers I could find (add_pa_data_element(), add_padata(), grow_pa_list()) put the list first.

Second, as a function which claims caller-provided memory rather than copying, there is an argument for also claiming the memory on failure (by freeing it).  That contract is easier to use, and is more consistent with how a Rust function would work: if you pass an object to a function (not a reference), that function now owns the object and the caller is done with it.  However, the last time I used that pattern, I found that Robbie's intuition as a C programmer didn't match mine, and I wound up adding a comment to every call site (this was for the KDC's add_pa_data_element()).

This is the prior art I found:
* kdc_util.c add_pa_data_element(): adds a caller-constructed krb5_pa_data, claims memory on success or failure.  Accepting (pa-type, contents, length) would simplify many call sites, but a few call sites have pre-built krb5_pa_data values that they got from kdcpreauth modules.
* get_in_tkt.c add_padata(): copies caller-provided contents rather than claiming them.  For one call site this copy isn't really wanted, but there is one call site where it is.
* preauth2.c grow_pa_list(): adds a caller-provided array of pa-data values without copying, because one of the call sites gets that kind of input from a clpreauth module.  Doesn't claim memory on failure.

Having the helper function make a copy is the simplest to analyze, but is of course wasteful when we just constructed the value via an ASN.1 encode operation.  Conditionally copying based on a flag tends to confuse static analyzers.

Reply:
I think the problem for me was that there wasn't anything different in the function interface that distinguished it from one which didn't consume - the Rust equivalent more clearly captures this in the typesystem.  While I appreciate the consideration, I don't think we should abandon the idiom.  If we wanted to improve clarity, there are a couple things we could do - one would be passing a pointer to the object so that we can NULL it after freeing when we pass ownership, another might be to extend the existing `varname_in` and `varname_out` convention wither another suffix, ... or perhaps just unifying the list mutation functions to all work this way would be sufficient.

Reply:
I was too surprised that it claims memory on failure, but okay for claiming on success, I guess I don't expect much from a failed call.